File: copy_bench.js
const { performance } = require('perf_hooks')

function seedArray(arrsize, numsize = 100) {
    let seeded = [];
    for(let i = 0; i < arrsize; i++) seeded.push(Math.abs(Math.round(Math.random()*numsize)))
    return seeded;
}

function runMeasurement(a) {
    const t1 = performance.now()
    const copy1 = a.map(_ => _)
    const t2 = performance.now()
    console.log(`\tmap:\t${(t2 - t1) * 1000}\tmicroseconds`, copy1.length)
    
    const t3 = performance.now()
    const copy2 = [...a]
    const t4 = performance.now()
    console.log(`\t...:\t${(t4 - t3) * 1000}\tmicroseconds`, copy2.length)
    
    const t5 = performance.now()
    const copy3 = [];
    for(const i of a) copy3.push(i)
    const t6 = performance.now()
    console.log(`\tfor-of:\t${(t6 - t5) * 1000}\tmicroseconds`, copy3.length)
    
    const t7 = performance.now()
    const copy4 = [];
    for(let i = 0; i<a.length; i++) copy4[i] = a[i];
    const t8 = performance.now()
    console.log(`\tfor:\t${(t8 - t7) * 1000}\tmicroseconds`, copy4.length)
    
    const t9 = performance.now()
    const copy5 = [];
    for(const i in a) copy5[i] = a[i];
    const t10 = performance.now()
    console.log(`\tfor-in:\t${(t10 - t9) * 1000}\tmicroseconds`, copy5.length)
}


function main() {
    const sizes = [10, 100, 1e3, 1e6, 1e7]
    for (let size of sizes) {
        console.log("Running Measure for size: ", size.toExponential())
        runMeasurement(seedArray(size))
    }
}

main()

File: copy_bench.cpp
// clang++ -std=c++2b copy_bench.cpp -o copy_bench
#include <iostream>
#include <vector>
#include <chrono>

using namespace std;

vector<int> seedArray(int arrsize, int numsize = 100) {
    vector<int> seeded;
    for (int i = 0; i < arrsize; i++) {
        seeded.push_back(rand() % numsize);
    }
    return seeded;
}

void runMeasurement(vector<int> a) {
    auto t1 = chrono::high_resolution_clock::now();
    vector<int> copy1 = a;
    auto t2 = chrono::high_resolution_clock::now();
    cout << "\tvector:\t" << chrono::duration_cast<chrono::microseconds>(t2 - t1).count() << "\tmicroseconds, " << copy1.size() << endl;

    auto t3 = chrono::high_resolution_clock::now();
    vector<int> copy2(a);
    auto t4 = chrono::high_resolution_clock::now();
    cout << "\tcopy:\t" << chrono::duration_cast<chrono::microseconds>(t4 - t3).count() << "\tmicroseconds, " << copy2.size() << endl;

    auto t5 = chrono::high_resolution_clock::now();
    vector<int> copy3;
    for (const auto& i : a) {
        copy3.push_back(i);
    }
    auto t6 = chrono::high_resolution_clock::now();
    cout << "\trange-for:\t" << chrono::duration_cast<chrono::microseconds>(t6 - t5).count() << "\tmicroseconds, " << copy3.size() << endl;

    auto t7 = chrono::high_resolution_clock::now();
    vector<int> copy4;
    for (size_t i = 0; i < a.size(); i++) {
        copy4.push_back(a[i]);
    }
    auto t8 = chrono::high_resolution_clock::now();
    cout << "\tfor:\t" << chrono::duration_cast<chrono::microseconds>(t8 - t7).count() << "\tmicroseconds, " << copy4.size() << endl;

    auto t9 = chrono::high_resolution_clock::now();
    vector<int> copy5;
    for (auto it = a.begin(); it != a.end(); ++it) {
        copy5.push_back(*it);
    }
    auto t10 = chrono::high_resolution_clock::now();
    cout << "\titerator:\t" << chrono::duration_cast<chrono::microseconds>(t10 - t9).count() << "\tmicroseconds, " << copy5.size() << endl;
}

void mainFunction() {
    srand(time(NULL));
    vector<int> sizes = {10, 100, 1000, 1000000, 10000000};
    for (auto size : sizes) {
        cout << "Running Measure for size: " << size << endl;
        vector<int> a = seedArray(size);
        runMeasurement(a);
    }
}

int main() {
    mainFunction();
    return 0;
}
